File: programmable_video/ios/Classes/RemoteParticipantListener.swift
// swiftlint:disable type_body_length
import Flutter
import Foundation
import TwilioVideo

class RemoteParticipantListener: BaseListener, RemoteParticipantDelegate {
    let TAG = "RemoteParticipantListener"

    func remoteParticipantDidDisableAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        debug("remoteParticipantDidDisableAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("audioTrackDisabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidEnableAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        debug("remoteParticipantDidEnableAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("audioTrackEnabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidPublishAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        debug("remoteParticipantDidPublishAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackPublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidUnpublishAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        debug("remoteParticipantDidUnpublishAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackUnpublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func didSubscribeToAudioTrack(audioTrack: RemoteAudioTrack, publication: RemoteAudioTrackPublication, participant: RemoteParticipant) {
        debug("didSubscribeToAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackSubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func didFailToSubscribeToAudioTrack(publication: RemoteAudioTrackPublication, error: Error, participant: RemoteParticipant) {
        debug("didFailToSubscribeToAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackSubscribedFailed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ], error: error)
    }

    func didUnsubscribeFromAudioTrack(audioTrack: RemoteAudioTrack, publication: RemoteAudioTrackPublication, participant: RemoteParticipant) {
        debug("didUnsubscribeFromAudioTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackUnsubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication),
            "remoteAudioTrack": RemoteParticipantListener.remoteAudioTrackToDict(audioTrack) as Any
        ])
    }

    func remoteParticipantDidPublishDataTrack(participant: RemoteParticipant, publication: RemoteDataTrackPublication) {
        debug("remoteParticipantDidPublishDataTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("dataTrackPublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteDataTrackPublication": RemoteParticipantListener.remoteDataTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidUnpublishDataTrack(participant: RemoteParticipant, publication: RemoteDataTrackPublication) {
        debug("remoteParticipantDidUnpublishDataTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("dataTrackUnpublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteDataTrackPublication": RemoteParticipantListener.remoteDataTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantNetworkQualityLevelDidChange(participant: RemoteParticipant, networkQualityLevel: NetworkQualityLevel) {
        debug("didChangeNetworkQualityLevel =>"  +
                                                    "sid: \(participant.sid ?? "")")
        sendEvent("networkQualityLevelChanged", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "networkQualityLevel": Mapper.networkQualityLevelToString(networkQualityLevel) as Any
        ])
    }

    func didSubscribeToDataTrack(dataTrack: RemoteDataTrack, publication: RemoteDataTrackPublication, participant: RemoteParticipant) {
        debug("didSubscribeToDataTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("dataTrackSubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteDataTrackPublication": RemoteParticipantListener.remoteDataTrackPublicationToDict(publication)
        ])
        publication.remoteTrack?.delegate = SwiftTwilioProgrammableVideoPlugin.remoteDataTrackListener
    }

    func didFailToSubscribeToDataTrack(publication: RemoteDataTrackPublication, error: Error, participant: RemoteParticipant) {
        debug("didFailToSubscribeToDataTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("dataTrackSubscriptionFailed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteDataTrackPublication": RemoteParticipantListener.remoteDataTrackPublicationToDict(publication)
        ], error: error)
    }

    func didUnsubscribeFromDataTrack(dataTrack: RemoteDataTrack, publication: RemoteDataTrackPublication, participant: RemoteParticipant) {
        debug("remoteParticipantDidUnpublishDataTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("dataTrackUnsubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteDataTrackPublication": RemoteParticipantListener.remoteDataTrackPublicationToDict(publication),
            "remoteDataTrack": RemoteDataTrackListener.remoteDataTrackToDict(dataTrack)
        ])
    }

    func remoteParticipantDidDisableVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        debug("remoteParticipantDidDisableVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("videoTrackDisabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidEnableVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        debug("remoteParticipantDidEnableVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("videoTrackEnabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidPublishVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        debug("remoteParticipantDidPublishVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackPublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidUnpublishVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        debug("remoteParticipantDidUnpublishVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackUnpublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func didSubscribeToVideoTrack(videoTrack: RemoteVideoTrack, publication: RemoteVideoTrackPublication, participant: RemoteParticipant) {
        debug("didSubscribeToVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackSubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func didFailToSubscribeToVideoTrack(publication: RemoteVideoTrackPublication, error: Error, participant: RemoteParticipant) {
        debug("didFailToSubscribeToVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackSubscribedFailed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ], error: error)
    }

    func didUnsubscribeFromVideoTrack(videoTrack: RemoteVideoTrack, publication: RemoteVideoTrackPublication, participant: RemoteParticipant) {
        debug("didUnsubscribeFromVideoTrack => " +
            "trackSid: \(publication.trackSid), " +
            "isTrackEnabled: \(publication.isTrackEnabled), " +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackUnsubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication),
            "remoteVideoTrack": RemoteParticipantListener.remoteVideoTrackToDict(videoTrack) as Any
        ])
    }

    public static func remoteParticipantToDict(_ remoteParticipant: RemoteParticipant, noTracks: Bool = false) -> [String: Any] {
        let remoteAudioTrackPublications: Any? = noTracks ? nil : remoteParticipant.remoteAudioTracks.map { (it) -> [String: Any] in
            return RemoteParticipantListener.remoteAudioTrackPublicationToDict(it)
        }
        let remoteVideoTrackPublications: Any? = noTracks ? nil : remoteParticipant.remoteVideoTracks.map({ (it) -> [String: Any] in
            return remoteVideoTrackPublicationToDict(it)
        })

        return [
            "identity": remoteParticipant.identity,
            "sid": remoteParticipant.sid as Any,
            "remoteAudioTrackPublications": remoteAudioTrackPublications as Any,
            "remoteVideoTrackPublications": remoteVideoTrackPublications as Any
        ]
    }

    static func remoteAudioTrackPublicationToDict(_ remoteAudioTrackPublication: RemoteAudioTrackPublication) -> [String: Any] {
        return [
            "sid": remoteAudioTrackPublication.trackSid,
            "name": remoteAudioTrackPublication.trackName,
            "enabled": remoteAudioTrackPublication.isTrackEnabled,
            "subscribed": remoteAudioTrackPublication.isTrackSubscribed,
            "remoteAudioTrack": remoteAudioTrackToDict(remoteAudioTrackPublication.remoteTrack) as Any
        ]
    }

    static func remoteAudioTrackToDict(_ remoteAudioTrack: RemoteAudioTrack?) -> [String: Any]? {
        if let remoteAudioTrack = remoteAudioTrack {
            return [
                "sid": remoteAudioTrack.sid,
                "name": remoteAudioTrack.name,
                "enabled": remoteAudioTrack.isEnabled
            ]
        }
        return nil
    }

    static func remoteDataTrackPublicationToDict(_ remoteDataTrackPublication: RemoteDataTrackPublication) -> [String: Any] {
        return [
            "sid": remoteDataTrackPublication.trackSid,
            "name": remoteDataTrackPublication.trackName,
            "enabled": remoteDataTrackPublication.isTrackEnabled,
            "subscribed": remoteDataTrackPublication.isTrackSubscribed,
            "remoteDataTrack": RemoteDataTrackListener.remoteDataTrackToDict(remoteDataTrackPublication.remoteTrack) as Any
        ]
    }

    static func remoteVideoTrackPublicationToDict(_ remoteVideoTrackPublication: RemoteVideoTrackPublication) -> [String: Any] {
        return [
            "sid": remoteVideoTrackPublication.trackSid,
            "name": remoteVideoTrackPublication.trackName,
            "enabled": remoteVideoTrackPublication.isTrackEnabled,
            "subscribed": remoteVideoTrackPublication.isTrackSubscribed,
            "remoteVideoTrack": remoteVideoTrackToDict(remoteVideoTrackPublication.remoteTrack) as Any
        ]
    }

    static func remoteVideoTrackToDict(_ remoteVideoTrack: RemoteVideoTrack?) -> [String: Any]? {
        if let remoteAudioTrack = remoteVideoTrack {
            return [
                "sid": remoteAudioTrack.sid,
                "name": remoteAudioTrack.name,
                "enabled": remoteAudioTrack.isEnabled
            ]
        }
        return nil
    }

    func debug(_ msg: String) {
        SwiftTwilioProgrammableVideoPlugin.debug("\(TAG)::\(msg)")
    }
}

File: programmable_video/android/src/main/kotlin/twilio/flutter/twilio_programmable_video/RemoteParticipantListener.kt
package twilio.flutter.twilio_programmable_video

import com.twilio.video.NetworkQualityLevel
import com.twilio.video.RemoteAudioTrack
import com.twilio.video.RemoteAudioTrackPublication
import com.twilio.video.RemoteDataTrack
import com.twilio.video.RemoteDataTrackPublication
import com.twilio.video.RemoteParticipant
import com.twilio.video.RemoteVideoTrack
import com.twilio.video.RemoteVideoTrackPublication
import com.twilio.video.TwilioException

class RemoteParticipantListener : BaseListener(), RemoteParticipant.Listener {
    private val TAG = "RemoteParticipantListener"

    override fun onAudioTrackDisabled(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        debug("onAudioTrackDisabled => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}")

        sendEvent("audioTrackDisabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackEnabled(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        debug("onAudioTrackEnabled => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}")

        sendEvent("audioTrackEnabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackPublished(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        debug("onAudioTrackPublished => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackPublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackSubscribed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, remoteAudioTrack: RemoteAudioTrack) {
        debug("onAudioTrackSubscribed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackSubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, twilioException: TwilioException) {
        debug("onAudioTrackSubscriptionFailed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackSubscriptionFailed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ), twilioException)
    }

    override fun onAudioTrackUnpublished(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        debug("onAudioTrackUnpublished => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackUnpublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, remoteAudioTrack: RemoteAudioTrack) {
        debug("onAudioTrackUnsubscribed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackUnsubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication),
                "remoteAudioTrack" to remoteAudioTrackToMap(remoteAudioTrack)
        ))
    }

    override fun onDataTrackPublished(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication) {
        debug("onDataTrackPublished => " +
                "trackSid: ${remoteDataTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteDataTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteDataTrackPublication.isTrackSubscribed}")

        sendEvent("dataTrackPublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteDataTrackPublication" to remoteDataTrackPublicationToMap(remoteDataTrackPublication)
        ))
    }

    override fun onDataTrackSubscribed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, remoteDataTrack: RemoteDataTrack) {
        debug("onDataTrackSubscribed => " +
                "trackSid: ${remoteDataTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteDataTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteDataTrackPublication.isTrackSubscribed}")

        sendEvent("dataTrackSubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteDataTrackPublication" to remoteDataTrackPublicationToMap(remoteDataTrackPublication)
        ))
        remoteDataTrackPublication.remoteDataTrack?.setListener(TwilioProgrammableVideoPlugin.remoteDataTrackListener)
    }

    override fun onDataTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, twilioException: TwilioException) {
        debug("onDataTrackSubscriptionFailed => " +
                "trackSid: ${remoteDataTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteDataTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteDataTrackPublication.isTrackSubscribed}")

        sendEvent("dataTrackSubscriptionFailed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteDataTrackPublication" to remoteDataTrackPublicationToMap(remoteDataTrackPublication)
        ), twilioException)
    }

    override fun onDataTrackUnpublished(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication) {
        debug("onDataTrackUnpublished => " +
                "trackSid: ${remoteDataTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteDataTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteDataTrackPublication.isTrackSubscribed}")

        sendEvent("dataTrackUnpublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteDataTrackPublication" to remoteDataTrackPublicationToMap(remoteDataTrackPublication)
        ))
    }

    override fun onDataTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, remoteDataTrack: RemoteDataTrack) {
        debug("onDataTrackUnsubscribed => " +
                "trackSid: ${remoteDataTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteDataTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteDataTrackPublication.isTrackSubscribed}")

        sendEvent("dataTrackUnsubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteDataTrackPublication" to remoteDataTrackPublicationToMap(remoteDataTrackPublication),
                "remoteDataTrack" to RemoteDataTrackListener.remoteDataTrackToMap(remoteDataTrack)
        ))
    }

    override fun onNetworkQualityLevelChanged(remoteParticipant: RemoteParticipant, networkQualityLevel: NetworkQualityLevel) {
        debug("onNetworkQualityLevelChanged => sid: ${remoteParticipant.sid}")
        sendEvent("networkQualityLevelChanged", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "networkQualityLevel" to networkQualityLevel.toString()
        ))
    }

    override fun onVideoTrackDisabled(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        debug("onVideoTrackDisabled => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}")

        sendEvent("videoTrackDisabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackEnabled(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        debug("onVideoTrackEnabled => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}")

        sendEvent("videoTrackEnabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackPublished(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        debug("onVideoTrackPublished => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackPublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackSubscribed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, remoteVideoTrack: RemoteVideoTrack) {
        debug("onVideoTrackSubscribed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackSubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, twilioException: TwilioException) {
        debug("onVideoTrackSubscriptionFailed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackSubscriptionFailed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ), twilioException)
    }

    override fun onVideoTrackUnpublished(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        debug("onVideoTrackUnpublished => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackUnpublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, remoteVideoTrack: RemoteVideoTrack) {
        debug("onVideoTrackUnsubscribed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackUnsubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication),
                "remoteVideoTrack" to remoteVideoTrackToMap(remoteVideoTrack)
        ))
    }

    companion object {
        @JvmStatic
        fun remoteParticipantToMap(remoteParticipant: RemoteParticipant, noTracks: Boolean = false): Map<String, Any?> {
            val remoteAudioTrackPublications = if (!noTracks) remoteParticipant.remoteAudioTracks.map { remoteAudioTrackPublicationToMap(it) } else null
            val remoteVideoTrackPublications = if (!noTracks) remoteParticipant.remoteVideoTracks.map { remoteVideoTrackPublicationToMap(it) } else null

            return mapOf(
                    "identity" to remoteParticipant.identity,
                    "sid" to remoteParticipant.sid,
                    "networkQualityLevel" to remoteParticipant.networkQualityLevel.toString(),
                    "remoteAudioTrackPublications" to remoteAudioTrackPublications,
                    "remoteVideoTrackPublications" to remoteVideoTrackPublications
            )
        }

        @JvmStatic
        fun remoteAudioTrackPublicationToMap(remoteAudioTrackPublication: RemoteAudioTrackPublication): Map<String, Any?> {
            return mapOf(
                    "sid" to remoteAudioTrackPublication.trackSid,
                    "name" to remoteAudioTrackPublication.trackName,
                    "enabled" to remoteAudioTrackPublication.isTrackEnabled,
                    "subscribed" to remoteAudioTrackPublication.isTrackSubscribed,
                    "remoteAudioTrack" to remoteAudioTrackToMap(remoteAudioTrackPublication.remoteAudioTrack)
            )
        }

        @JvmStatic
        fun remoteAudioTrackToMap(remoteAudioTrack: RemoteAudioTrack?): Map<String, Any>? {
            if (remoteAudioTrack != null) {
                return mapOf(
                        "sid" to remoteAudioTrack.sid,
                        "name" to remoteAudioTrack.name,
                        "enabled" to remoteAudioTrack.isEnabled
                )
            }
            return null
        }

        @JvmStatic
        fun remoteDataTrackPublicationToMap(remoteDataTrackPublication: RemoteDataTrackPublication): Map<String, Any?> {
            return mapOf(
                    "sid" to remoteDataTrackPublication.trackSid,
                    "name" to remoteDataTrackPublication.trackName,
                    "enabled" to remoteDataTrackPublication.isTrackEnabled,
                    "subscribed" to remoteDataTrackPublication.isTrackSubscribed,
                    "remoteDataTrack" to RemoteDataTrackListener.remoteDataTrackToMap(remoteDataTrackPublication.remoteDataTrack)
            )
        }

        @JvmStatic
        fun remoteVideoTrackPublicationToMap(remoteVideoTrackPublication: RemoteVideoTrackPublication): Map<String, Any?> {
            return mapOf(
                    "sid" to remoteVideoTrackPublication.trackSid,
                    "name" to remoteVideoTrackPublication.trackName,
                    "enabled" to remoteVideoTrackPublication.isTrackEnabled,
                    "subscribed" to remoteVideoTrackPublication.isTrackSubscribed,
                    "remoteVideoTrack" to remoteVideoTrackToMap(remoteVideoTrackPublication.remoteVideoTrack)
            )
        }

        @JvmStatic
        fun remoteVideoTrackToMap(remoteVideoTrack: RemoteVideoTrack?): Map<String, Any>? {
            if (remoteVideoTrack != null) {
                return mapOf(
                        "sid" to remoteVideoTrack.sid,
                        "name" to remoteVideoTrack.name,
                        "enabled" to remoteVideoTrack.isEnabled
                )
            }
            return null
        }
    }

    internal fun debug(msg: String) {
        TwilioProgrammableVideoPlugin.debug("$TAG::$msg")
    }
}
